agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/StartupDiagnostics.java

```diff
@@ -0,0 +1,182 @@
+  //    ],
+  //    c2: {
+  //      Exclude: true
+  //    }
+  //  }
+  // ]
+  private static final String APPPLICATIONINSIGHTS_DEBUG_JIT_C_2_DISABLING_ENABLED =
+      "appplicationinsights.debug.jit.c2-disabling.enabled";
+
+  private static final String APPPLICATIONINSIGHTS_DEBUG_JVM_FLAGS_ENABLED =
+      "appplicationinsights.debug.jvm-flags.enabled";
+
+  // Execute with -XX:NativeMemoryTracking=summary
+  private static final String APPPLICATIONINSIGHTS_DEBUG_NATIVE_MEM_TRACKING_ENABLED =
+      "appplicationinsights.debug.native-mem-tracking.enabled";
+
+  // "file" (default value) / "console" / "file-console"
+  public static final String APPPLICATIONINSIGHTS_DEBUG_DIAG_EXPORT =
+      "appplicationinsights.debug.diag-export";
+  private final Logger startupLogger;
+
+  public StartupDiagnostics(Logger startupLogger) {
+    this.startupLogger = startupLogger;
+  }
+
+  static class DiagnosticsReport {
+
+    StringBuilder diagnosticBuilder = new StringBuilder();
+
+    void addDiagnostic(String diagnostic) {
+      diagnosticBuilder.append(diagnostic);
+      diagnosticBuilder.append(System.lineSeparator());
+      diagnosticBuilder.append(System.lineSeparator());
+    }
+
+    boolean isEmpty() {
+      return diagnosticBuilder.length() == 0;
+    }
+
+    @Override
+    public String toString() {
+      return diagnosticBuilder.toString();
+    }
+  }
+
+  void execute() {
+
+    executeDiagsAndGenerateReport();
+
+    if (Boolean.getBoolean(APPPLICATIONINSIGHTS_DEBUG_JIT_C_2_DISABLING_ENABLED)) {
+      disableC2CompilerDirectives();
+    }
+  }
+
+  private void executeDiagsAndGenerateReport() {
+    DiagnosticsReport diagnosticsReport = new DiagnosticsReport();
+
+    if (Boolean.getBoolean("rss")) {
+      String os = System.getProperty("os.name");
+      if (os.equals("Linux")) {
+        String residentSetSize = findResidentSetSize();
+        diagnosticsReport.addDiagnostic(residentSetSize);
+      }
+    }
+
+    if (Boolean.getBoolean(APPPLICATIONINSIGHTS_DEBUG_JVM_FLAGS_ENABLED)) {
+      String jvmFlags = executeJvmFlagsDiag();
+      diagnosticsReport.addDiagnostic(jvmFlags);
+    }
+
+    if (Boolean.getBoolean(APPPLICATIONINSIGHTS_DEBUG_NATIVE_MEM_TRACKING_ENABLED)) {
+      String nativeSummary = executeNativeMemoryDiag();
+      diagnosticsReport.addDiagnostic(nativeSummary);
+    }
+
+    generateReport(diagnosticsReport);
+  }
+
+  private void generateReport(DiagnosticsReport diagnosticsReport) {
+    if (!diagnosticsReport.isEmpty()) {
+      String diagExport = System.getProperty(APPPLICATIONINSIGHTS_DEBUG_DIAG_EXPORT);
+      if ("console".equals(diagExport) || "file-console".equals(diagExport)) {
+        startupLogger.info("Start-up diagnostics" + File.separator + diagnosticsReport);
+      }
+      // "file" (default value) / "console" / "file-console"
+      if (diagExport == null || "file".equals(diagExport) || "file-console".equals(diagExport)) {
+        saveIntoFile(diagnosticsReport);
+      }
```

Comment:
I'd suggest that we just write it to the normal application insights self-diagnostics stream (which is by default console and applicationinsights.log), since that output is already configurable (https://docs.microsoft.com/en-us/azure/azure-monitor/app/java-standalone-config#self-diagnostics)

we could even enable all of these reports automatically when self-diagnostic level is `debug`, similar to how we dump all jmx metrics at that level.